runtime/src/bank.rs

```diff
@@ -369,91 +386,99 @@ impl Bank {
     /// Create a new bank that points to an immutable checkpoint of another bank.
     pub fn new_from_parent(parent: &Arc<Bank>, collector_id: &Pubkey, slot: Slot) -> Self {
         parent.freeze();
-        assert_ne!(slot, parent.slot());
 
+        let config = parent.bank_config();
+        let bank = Self::new_from_config(config, parent.hash(), slot);
+
+        // inherit shared (resource-like) fields inside the validator process
         let rc = BankRc {
             accounts: Arc::new(Accounts::new_from_parent(
                 &parent.rc.accounts,
                 slot,
-                parent.slot(),
+                bank.parent_slot,
             )),
             parent: RwLock::new(Some(parent.clone())),
             slot,
         };
         let src = StatusCacheRc {
             status_cache: parent.src.status_cache.clone(),
         };
-        let epoch_schedule = parent.epoch_schedule;
-        let epoch = epoch_schedule.get_epoch(slot);
+        bank.collector_id = *collector_id;
+        bank.genesis_creation_time = parent.genesis_creation_time;
+        bank.entered_epoch_callback = parent.entered_epoch_callback.clone();
+        bank.blockhash_queue = RwLock::new(parent.blockhash_queue.read().unwrap().clone());
+        bank.last_vote_sync: AtomicU64::new(parent.last_vote_sync.load(Ordering::Relaxed));
+        bank.message_processor = MessageProcessor::default();
+
+        // TODO: create those from AccountsDB? is it possible?
+        // or verify structural integrity of these caches from snapshot instead of hashing into the slot hash
+        //   is it possible first of all with tamper-residence?
+        // due to possibility of being large in number for these fields, is there perf.-wise concerns?
+        // this is like verify_hash_internal_state
+        stakes: RwLock::new(parent.stakes.read().unwrap().clone_with_epoch(epoch)),
+        epoch_stakes: parent.epoch_stakes.clone(),
+        storage_accounts: RwLock::new(parent.storage_accounts.read().unwrap().clone()),
+
+        assert_ne!(bank.slot, bank.parent_slot);
+
+        datapoint_debug!(
+            "bank-new_from_parent-heights",
+            ("slot_height", bank.slot, i64),
+            ("block_height", bank.block_height, i64)
+        );
+
+        bank
+    }
+
+    // derive various carried-over states from the BankConfig of previous bank!
+    // this is potentially untrusted which should be protected by hashing and SPV
+    pub fn new_from_config(config, parent_hash, slot) -> Self {
+        let epoch = config.epoch_schedule.get_epoch(slot);
 
         let mut new = Bank {
-            rc,
-            src,
             slot,
             epoch,
-            blockhash_queue: RwLock::new(parent.blockhash_queue.read().unwrap().clone()),
-
-            // TODO: clean this up, soo much special-case copying...
```

Comment:
Phew, this initialization code block holds quite a lot fields. ;)

I've examined each fields for BankConfig or not; As this is still a draft, the categorization is rough and some might be wrong.